typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -511,15 +517,42 @@
        ;; TODO: this is not quite right for case->
        [(Prompt-Tagof: s (Fun: (list (Arrow: ts _ _ _))))
         (prompt-tag/sc (map t->sc ts) (t->sc s))]
-       [(F: v) #:when (string-prefix? (symbol->string v) "self-")
-               (fail #:reason "contract generation not supported for Self")]
-       ;; TODO
+       [(Exist: (list n) (Fun: (list li-arrs ...)))
+        (define (occur? t)
+          (if (or (not t) (empty? t)) #f
+              (set-member? (free-vars-names (free-vars* t)) n)))
+
+        (define (err) (fail #:reason "contract generation only supports Exist Type in this form: (Exist (X) (-> ty1 .. (-> X ty ... ty2) : X)"))
+        
+        (define (get-dom-rng li-arrs)
+          (match li-arrs
+            [(list (Arrow: (list dom) rst kw
+                           (Values: (list (Result: (and (Fun: (list (Arrow: (list-rest (F: n1) a ... _) rst_i kw_i r))) rng)
+                                                   (PropSet: (TypeProp: _ (F: n1)) _) _)))))
+             #:when (and (not (ormap occur? (list rst kw rst_i kw_i)))
+                         (eq? n1 n))
+             (values dom rng)]
+            [_ (err)]))
+        
+        (define-values (dom rng) (get-dom-rng li-arrs))
+        (define/with-syntax name n)
+        (define lhs (t->sc/neg dom))
+        (define eq-name (flat/sc #'(eq/c name)))
```

Comment:
`eq/c` is probably much slower than `(lambda (x) (eq? x name))` ... do we really need `make-contract`?

Reply:
we probably need make-contract to make related error messages readable.

Reply:
does this file still need to import `racket/contract/combinator` ?

if so, please move the require up with the others (from line 54)